File: dlang/psi-impl/src/main/java/io/github/intellij/dlanguage/stubs/types/DLanguageNamedImportBindStubElementType.java
package io.github.intellij.dlanguage.stubs.types;

import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import io.github.intellij.dlanguage.psi.DLanguageNamedImportBind;
import io.github.intellij.dlanguage.psi.impl.named.DLanguageNamedImportBindImpl;
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes;
import io.github.intellij.dlanguage.stubs.DlangNamedImportBindStub;
import java.io.IOException;
import org.jetbrains.annotations.NotNull;

public class DLanguageNamedImportBindStubElementType extends DNamedStubElementType<DlangNamedImportBindStub, DLanguageNamedImportBind> {
    public DLanguageNamedImportBindStubElementType(final String debugName) {
        super(debugName);
    }

    @Override
    public DLanguageNamedImportBind createPsi(@NotNull final DlangNamedImportBindStub stub) {
        return new DLanguageNamedImportBindImpl(stub, this);
    }

    @NotNull
    @Override
    public DlangNamedImportBindStub createStub(@NotNull final DLanguageNamedImportBind psi, final StubElement parentStub) {
        return new DlangNamedImportBindStub(parentStub, this, psi.getName(), psi.getAttributes());
    }

    @Override
    public void serialize(@NotNull final DlangNamedImportBindStub stub, @NotNull final StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());
        stub.getAttributes().write(dataStream);
    }

    @NotNull
    @Override
    public DlangNamedImportBindStub deserialize(@NotNull final StubInputStream dataStream, final StubElement parentStub) throws IOException {
        return new DlangNamedImportBindStub(parentStub, this, dataStream.readName(),
            DAttributes.Companion.read(dataStream));
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/DLanguageNamedImportBindStubElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import io.github.intellij.dlanguage.psi.DLanguageNamedImportBind
import io.github.intellij.dlanguage.psi.impl.named.DLanguageNamedImportBindImpl
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes.Companion.read
import io.github.intellij.dlanguage.stubs.DLanguageNamedImportBindStub
import java.io.IOException

class DLanguageNamedImportBindStubElementType(debugName: String) :
    DNamedStubElementType<DLanguageNamedImportBindStub, DLanguageNamedImportBind>(debugName) {
    override fun createPsi(stub: DLanguageNamedImportBindStub): DLanguageNamedImportBind {
        return DLanguageNamedImportBindImpl(stub, this)
    }

    override fun createStub(psi: DLanguageNamedImportBind, parentStub: StubElement<*>?): DLanguageNamedImportBindStub {
        return DLanguageNamedImportBindStub(parentStub, this, psi.name, psi.attributes)
    }

    @Throws(IOException::class)
    override fun serialize(stub: DLanguageNamedImportBindStub, dataStream: StubOutputStream) {
        dataStream.writeName(stub.name)
        stub.attributes.write(dataStream)
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>?): DLanguageNamedImportBindStub {
        return DLanguageNamedImportBindStub(
            parentStub, this, dataStream.readName(),
            read(dataStream)
        )
    }
}
